ancestrymanager/ancestrymanager.go

```diff
@@ -1,189 +1,240 @@
 }
 
-// resourceAncestryManager provides common methods for retrieving ancestry from resources
-type resourceAncestryManager struct {
+type manager struct {
+	// The logger.
 	errorLogger *zap.Logger
+	// GCP resource manager service. If this field is nil, online lookups will .
+	// be disabled.
+	resourceManager *cloudresourcemanager.Service
+	// Cache to prevent multiple network calls for looking up the same
+	// resource's ancestry. The map key is the resource itself, in the format of
+	// "<type>/<id>", ancestors are sorted from closest to furthest.
+	ancestorCache map[string][]string
 }
 
-func (m *resourceAncestryManager) getFolderAncestry(folderID string) ([]*cloudresourcemanager.Ancestor, error) {
-	// TODO(morgantep): Incorporate folders.GetAncestry from v2alpha1 API
-	m.errorLogger.Info(fmt.Sprintf("Retrieving ancestry for folder: %s", folderID))
-
-	return []*cloudresourcemanager.Ancestor{
-		&cloudresourcemanager.Ancestor{
-			ResourceId: &cloudresourcemanager.ResourceId{
-				Type: "folder",
-				Id:   folderID,
-			},
-		},
-		&cloudresourcemanager.Ancestor{
-			ResourceId: &cloudresourcemanager.ResourceId{
-				Type: "organization",
-				Id:   "unknown",
-			},
-		},
-	}, nil
+// New returns AncestryManager that can be used to fetch ancestry information
+// for a project. entries takes project as key and ancestry as value to
+// pre-warm the offline cache. If no resourceManager is provided, API requests
+// for ancestry will be disabled.
+func New(resourceManager *cloudresourcemanager.Service, entries map[string]string, errorLogger *zap.Logger) (AncestryManager, error) {
+	return newManager(resourceManager, entries, errorLogger)
 }
 
-func (m *resourceAncestryManager) getAncestryFromResource(tfData resources.TerraformResourceData, cai resources.Asset) ([]*cloudresourcemanager.Ancestor, bool) {
-	m.errorLogger.Info(fmt.Sprintf("Retrieving ancestry from resource (type=%s)", cai.Type))
+func newManager(resourceManager *cloudresourcemanager.Service, entries map[string]string, errorLogger *zap.Logger) (*manager, error) {
+	am := &manager{
+		ancestorCache:   map[string][]string{},
+		errorLogger:     errorLogger,
+		resourceManager: resourceManager,
+	}
+	am.initAncestryCache(entries)
+	return am, nil
+}
 
-	switch cai.Type {
-	case "cloudresourcemanager.googleapis.com/Project", "cloudbilling.googleapis.com/ProjectBillingInfo":
-		// Prefer project number to project id if available;
-		// CAI exports use project number.
-		projectID, ok := tfData.GetOk("number")
-		if !ok {
-			projectID, ok = tfData.GetOk("project_id")
-			if !ok || projectID == "" {
-				return nil, false
+func (m *manager) initAncestryCache(entries map[string]string) {
+	for item, ancestry := range entries {
+		if item != "" && ancestry != "" {
+			ancestors, err := parseAncestryPath(ancestry)
+			if err != nil {
+				continue
+			}
+			if !strings.HasPrefix(item, "projects/") && !strings.HasPrefix(item, "folders/") {
```

Comment:
What problems could it introduce? If user provides invalid key, then the cache is invalid, then it will use cloud resource API to fetch ancestors.

Reply:
What should be the expected behavior? 
For current code, if it does not start with projects/ or folders/, then it is considered as a pure project name(mostly in that case, we do not need to fix the test cases), so no error will be surface. 
I think it can be less restrictive, to allow it to check whether it check project(s)/ and folder(s)/, but I am not sure whether we should surface if users put "orgs/", "organizations/" etc?

Reply:
My understanding is that a valid cache id is either:

- `folders/<number>`
- `projects/<number>`
- `projects/<id>` (at some point we might want to maintain a separate project id -> project number cache but that's separate work)

Currently if a user provided a value like `organizations/12345`, my read of the code is that it would be put in the cache as `projects/organizations/12345` - which wouldn't really hurt anything but isn't what the user intended.

A better example might be if a user provided a value like `foldres/12345` (typo). Currently that would end up in the cache as `projects/foldres/12345`, which is not what the user intended and could actually negatively impact validation behavior.

For the expected behavior - it would be great for invalid cache keys to be surfaced to the user as errors so they can be fixed - they're essentially input errors.

None of those cases support a `/` on the right-hand side, so one simple validation could be to make sure that if a `/` is present, the left side is a valid resource type.

Reply:
Done. I added check for both key and value. 
If any of them contains invalid resource type (not projects/, folders/, organizaitons/), init ancestry manager throws an error.